google/cloud/spanner/examples/database_admin_snippets.cc

```diff
@@ -0,0 +1,180 @@
+  }
+
+  //! [drop-database] [START spanner_drop_database]
+  [](std::string const& project_id, std::string const& instance_id,
+     std::string const& database_id) {
+    google::cloud::spanner::DatabaseAdminClient client;
+    google::cloud::Status status =
+        client.DropDatabase(project_id, instance_id, database_id);
+    if (!status.ok()) {
+      throw std::runtime_error(status.message());
+    }
+    std::cout << "Database " << database_id << " successfully dropped\n";
+  }
+  //! [drop-database] [END spanner_drop_database]
+  (argv[0], argv[1], argv[2]);
+}
+
+int RunOneCommand(std::vector<std::string> argv) {
+  using CommandType = std::function<void(std::vector<std::string> const&)>;
+
+  std::map<std::string, CommandType> commands = {
+      {"create-database", &CreateDatabase},
+      {"drop-database", &DropDatabase},
+  };
+
+  std::string usage_msg = [&argv, &commands] {
+    auto last_slash = std::string(argv[0]).find_last_of("/\\");
+    auto program = argv[0].substr(last_slash + 1);
+    std::string usage;
+    usage += "Usage: ";
+    usage += program;
+    usage += " <command> [arguments]\n\n";
+    usage += "Commands:\n";
+    for (auto&& kv : commands) {
+      try {
+        kv.second({});
+      } catch (std::exception const& ex) {
+        usage += "    ";
+        usage += ex.what();
+        usage += "\n";
+      }
+    }
+    return usage;
+  }();
+
+  if (argv.size() < 2) {
+    std::cerr << "Missing command argument\n" << usage_msg << "\n";
+    return 1;
+  }
+  std::string command_name = argv[1];
+  argv.erase(argv.begin());  // remove the program name from the list.
+  argv.erase(argv.begin());  // remove the command name from the list.
+
+  auto command = commands.find(command_name);
+  if (commands.end() == command) {
+    std::cerr << "Unknown command " << command_name << "\n"
+              << usage_msg << "\n";
+    return 1;
+  }
+
+  // Run the command.
+  command->second(argv);
+  return 0;
+}
+
+void RunAll() {
+  auto project_id =
+      google::cloud::internal::GetEnv("GOOGLE_CLOUD_PROJECT").value_or("");
+  auto instance_id =
+      google::cloud::internal::GetEnv("GOOGLE_CLOUD_CPP_SPANNER_INSTANCE")
+          .value_or("");
+
+  if (project_id.empty()) {
+    throw std::runtime_error("GOOGLE_CLOUD_PROJECT is not set or is empty");
+  }
+  if (instance_id.empty()) {
+    throw std::runtime_error(
+        "GOOGLE_CLOUD_CPP_SPANNER_INSTANCE is not set or is empty");
+  }
+  auto generator = google::cloud::internal::MakeDefaultPRNG();
+  std::string database_id = RandomDatabaseName(generator);
+
+  RunOneCommand({"", "create-database", project_id, instance_id, database_id});
+  RunOneCommand({"", "drop-database", project_id, instance_id, database_id});
+}
+
+bool AutoRun() {
+  return google::cloud::internal::GetEnv("GOOGLE_CLOUD_CPP_AUTO_RUN_EXAMPLES")
+             .value_or("") == "yes";
+}
+
+}  // namespace
+
+int main(int ac, char* av[]) try {
```

Comment:
My understanding of [function try blocks](https://en.cppreference.com/w/cpp/language/function-try-block) is that they're intended to catch exceptions from member initializers in a constructor. They do work on any function (including `main()`), but I *think* for regular functions they're equivalent to the normal form of 
```cc
int main() {
  try {
    ...
  } catch (...) {
    ...
  }
}
``` 

I have a slight preference for using the vanilla try/catch block form for functions since I think it's more typical and less surprising. Of course, we can always feel free to use a function-try-block in a c'tor where it may be needed to catch  member initialization exceptions.

Thoughts?

Reply:
> They do work on any function (including `main()`), but I _think_ for regular functions they're equivalent to the normal form of

Agree.

> I have a slight preference for using the vanilla try/catch block form for functions

I have a slight preference to the function try block, because it saves on the extra indentation.

> since I think it's more typical and less surprising.

I do not find them surprising at all, specially around `main()`, but that might be just me.  As to "typical", I am not sure, in a previous life, when we used exceptions a lot (I mean, we did not *throw* them a lot, but we assumed they could happen anywhere) we wrote `main()` like this.

Seems like we have a tie on "slight preference", should we flip a coin?

Reply:
Keeping it sounds fine. Saving a little indentation is a fair point, and it's more concrete than my opinion about typical and surprising.

Maybe I'll grow to like it, like I've grown to like east const :-)